In [0]:
import pandas as pd
import numpy as np
import csv
import re
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
travelpony = pd.read_csv('drive/My Drive/Colab Notebooks/in class exercises/Travel Pony Facebook.csv')
travelpony.head

<bound method NDFrame.head of       Start Date  ... Cost per Page Engagement (USD)
0     2013-12-02  ...                       1.219512
1     2013-12-01  ...                       1.395349
2     2013-11-30  ...                       1.190476
3     2013-11-29  ...                       1.167315
4     2013-11-22  ...                       2.512857
...          ...  ...                            ...
3700  2013-06-11  ...                       1.050000
3701  2013-06-10  ...                       0.533333
3702  2013-06-09  ...                       0.990000
3703  2013-06-08  ...                       0.316667
3704  2013-06-07  ...                       0.000000

[3705 rows x 20 columns]>

In [0]:
list(travelpony)

['Start Date',
 'End Date',
 'Account',
 'Campaign Name',
 'Reach',
 'Frequency',
 'Impressions',
 'Clicks',
 'Unique Clicks',
 'Amount Spent (USD)',
 'Page Likes',
 'Page Engagement',
 'Post Engagement',
 'Post Likes',
 'Post Comments',
 'Post Shares',
 'Photo Views',
 'Website Clicks',
 'Cost per Page Like (USD)',
 'Cost per Page Engagement (USD)']

In [0]:
travelpony['Cost per Impression'] = travelpony['Amount Spent (USD)'] / travelpony['Impressions']
travelpony.columns

Index(['Start Date', 'End Date', 'Account', 'Campaign Name', 'Reach',
       'Frequency', 'Impressions', 'Clicks', 'Unique Clicks',
       'Amount Spent (USD)', 'Page Likes', 'Page Engagement',
       'Post Engagement', 'Post Likes', 'Post Comments', 'Post Shares',
       'Photo Views', 'Website Clicks', 'Cost per Page Like (USD)',
       'Cost per Page Engagement (USD)', 'Cost per Impression'],
      dtype='object')

In [0]:
travelpony['Cost per Impression']

0       0.000466
1       0.000652
2       0.000448
3       0.000452
4       0.002464
          ...   
3700    0.003512
3701    0.002936
3702    0.008919
3703    0.006643
3704    0.000439
Name: Cost per Impression, Length: 3705, dtype: float64

In [0]:
travelpony['Start Date'] = pd.to_datetime(travelpony['Start Date'])
travelpony['Day of Week'] = travelpony['Start Date'].dt.day_name()
travelpony.columns

Index(['Start Date', 'End Date', 'Account', 'Campaign Name', 'Reach',
       'Frequency', 'Impressions', 'Clicks', 'Unique Clicks',
       'Amount Spent (USD)', 'Page Likes', 'Page Engagement',
       'Post Engagement', 'Post Likes', 'Post Comments', 'Post Shares',
       'Photo Views', 'Website Clicks', 'Cost per Page Like (USD)',
       'Cost per Page Engagement (USD)', 'Cost per Impression', 'Day of Week'],
      dtype='object')

In [0]:
travelpony.groupby('Day of Week', as_index=False).agg({"Cost per Impression": "sum"})


Day of Week  Cost per Impression
0      Friday             2.056639
1      Monday             1.713976
2    Saturday             1.311720
3      Sunday             1.755447
4    Thursday             1.950631
5     Tuesday             1.440979
6   Wednesday             1.761665

**Number 1:**

According to the above summary, the day of week that works **best** (aka the *cheapest* to generate impressions) is **Saturday**.

The day of week that works **worst** (aka the *most expensive* to generate impressions) is **Friday**.


In [0]:
print(travelpony[['Amount Spent (USD)','Reach']].corr())

                    Amount Spent (USD)     Reach
Amount Spent (USD)            1.000000  0.703124
Reach                         0.703124  1.000000


In [0]:
print(travelpony[['Amount Spent (USD)','Frequency']].corr())

                    Amount Spent (USD)  Frequency
Amount Spent (USD)            1.000000   0.130201
Frequency                     0.130201   1.000000


In [0]:
print(travelpony[['Amount Spent (USD)','Unique Clicks']].corr())

                    Amount Spent (USD)  Unique Clicks
Amount Spent (USD)            1.000000       0.882993
Unique Clicks                 0.882993       1.000000


In [0]:
print(travelpony[['Amount Spent (USD)','Page Likes']].corr())

                    Amount Spent (USD)  Page Likes
Amount Spent (USD)            1.000000    0.757612
Page Likes                    0.757612    1.000000


**Number 2**: 

The correlation that is the strongest is the Amount Spent and the Unique Clicks with 0.883. This means that the more they spend, the more unique clicks they will get on a post.


In [0]:
import statsmodels.api as sm

In [0]:
independent_variable = travelpony[["Reach",'Frequency']]
dependent_variable = travelpony["Amount Spent (USD)"]

model = sm.OLS(dependent_variable, independent_variable).fit()
predictions = model.predict(independent_variable)

model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:     Amount Spent (USD)   R-squared (uncentered):                   0.541
Model:                            OLS   Adj. R-squared (uncentered):              0.541
Method:                 Least Squares   F-statistic:                              2187.
Date:                Wed, 13 Nov 2019   Prob (F-statistic):                        0.00
Time:                        00:07:44   Log-Likelihood:                         -16058.
No. Observations:                3705   AIC:                                  3.212e+04
Df Residuals:                    3703   BIC:                                  3.213e+04
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Reach          0.0019   3.19e-05     58.957      0.000       0.002       0.002
Frequency      4.7323      0.305     15.499      0.000       4.134       5.331
==============================================================================
Omnibus:                     2898.063   Durbin-Watson:                   1.351
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           404599.317
Skew:                           2.933   Prob(JB):                         0.00
Kurtosis:                      53.857   Cond. No.                     9.84e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.84e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Number 3:**

The variable that most strongly predicts unique clicks is Frequency. As Amount Spent increases by 1, the Reach increases by 0.0019 and the Frequency increases by 4.73. This means the more money spent, the higher the frequency will be. The R-squared is also fairly low, 0.541, meaning that this model explains 54.1% of the variance in the dependent variable (Amount Spent).